File: project1/src/task2.py
import json
from pyspark import SparkContext, SparkConf
import os
import sys
import logging
import time

def isValidFiles(review_file, business_file):
    if not os.path.isfile(review_file) or not os.path.isfile(business_file): 
        return False
    return True


def loadBusiness(jsonObj):
    categories = jsonObj.get("categories")
    categories_list = None
    if categories == None or categories == "":
        categories_list = []
    else:
        categories = categories.split(",")
        categories_list = [category.strip() for category in categories]
    return (jsonObj["business_id"], categories_list)


def loadReviews(jsonObj):
    return (jsonObj["business_id"], (float(jsonObj["stars"]), 1))


def mapCategory(values):
    categories_list = values[0]
    stars_count = values[1]
    categoriesMap = [(value, stars_count) for value in categories_list]
    return categoriesMap


def loadBusinessJson(business_file):
    id_cat = {}
    cat_count = {}
    with open(business_file, "r") as file:
        lines = file.readlines()

        for line in lines:
            jsonObj = json.loads(line)

            categories = jsonObj.get("categories")

            if categories != None and categories != "":
                categories = categories.split(",")
                categories_list = [category.strip() for category in categories]
                id_cat[jsonObj["business_id"]] = categories_list

                for category in categories_list:
                    if cat_count.get(category) == None:
                        cat_count[category] = [0.0, 0]
    return id_cat, cat_count


def mapBusinessReviews(review_file, id_cat, cat_count):
    file = open(review_file, "r")

    while True:
        line = file.readline()

        if not  line:
            break

        jsonObj = json.loads(line)

        categories_list = id_cat.get(jsonObj["business_id"])

        if categories_list != None:
            for category in categories_list:
                cat_count[category][0] += float(jsonObj["stars"])
                cat_count[category][1] += 1

    return cat_count


def get_top_avg(results, n_arg):
    avg_cat_dict = {}

    for pair in results:
        if avg_cat_dict.get(pair[1]) == None:
            avg_cat_dict[pair[1]] = [pair[0]]
        else:
            avg_cat_dict[pair[1]].append(pair[0])

    top_avg_stars = []

    for star, cateogories in avg_cat_dict.items():
        if n_arg == 0:
            break
        
        cateogories = sorted(cateogories)

        for category in cateogories:
            if n_arg == 0:
                break

            top_avg_stars.append([category, star])
            n_arg -= 1

    return top_avg_stars



def main(argv):
    assert len(argv) == 5, "Script takes 5 arguments"

    # Unpack arguments 
    review_file, business_file, output_file, if_spark, n_arg = argv

    # Check for valid file paths
    if not isValidFiles(review_file, business_file):
        print("Invalid path arguments")
        return

    n_arg = int(n_arg)
    out_dict = {}

    if if_spark == "spark":
        start = time.time()

        # Get Error Logs
        logging.getLogger("org").setLevel(logging.ERROR)

        conf = SparkConf().setMaster("local[*]").setAppName("Task2")
        sc = SparkContext(conf=conf).getOrCreate()

        # Create BusinessRDD and ReviewsRDD
        businessJson = sc.textFile(business_file)
        reviewJson = sc.textFile(review_file)

        businessRDD = businessJson.map(json.loads).map(lambda x: loadBusiness(x))
        reviewsRDD = reviewJson.map(json.loads).map(lambda x: loadReviews(x))

        # Natural Join Business and Reviews
        business_reviews = businessRDD.join(reviewsRDD).cache()

        category_stars = business_reviews.filter(lambda x: len(x[1][0]) != 0) \
                                         .map(lambda x: mapCategory(x[1])) \
                                         .flatMap(lambda x: x) \
                                         .reduceByKey(lambda x, y: (x[0] + y[0], x[1] + y[1])).cache()

        # Compute the average
        avg_stars = category_stars.mapValues(lambda x: x [0]/x[1]).sortBy(lambda x: x[1], ascending=False)

        results = avg_stars.collect()

        top_avg_stars = get_top_avg(results, n_arg)

        out_dict["result"] = top_avg_stars

        end = time.time()

        print("Task 2 ", end - start)

    else:
        start = time.time()

        id_cat, cat_count = loadBusinessJson(business_file)
        
        cat_count = mapBusinessReviews(review_file, id_cat, cat_count)

        del id_cat # free up memory

        avg_cat = {}

        for category, star_count in cat_count.items():
            try:
                avg_cat[category] = star_count[0]/star_count[1]
            except ZeroDivisionError as e:
                avg_cat[category] = 0.0

        del cat_count # free up memory

        # sort average catgory rating --> list of tuples
        avg_cat = sorted(avg_cat.items(), key=lambda kv: kv[1], reverse=True)
        
        top_avg_stars = get_top_avg(avg_cat, n_arg)

        out_dict["result"] = top_avg_stars

        end = time.time()

        print("Task 2 ", end - start)

    with open(output_file, "w") as file:
        json.dump(out_dict, file)
        

if __name__ == "__main__":
    main(sys.argv[1:])

File: project1_scala/hw1/src/main/scala/task2.scala
import java.io.FileWriter

import org.apache.log4j.{Level, Logger}
import org.apache.spark.SparkContext
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._

import Array._
import scala.collection.immutable.ListMap
import scala.collection.mutable.ArrayBuffer
import scala.io.Source
import scala.util.Sorting.quickSort
import scala.util.control.Breaks._

//noinspection DuplicatedCode
object task2 extends Serializable {

  def time[R](block: => R): R = {
    val t0 = System.nanoTime()
    val result = block
    val t1 = System.nanoTime()
    println("Elapsed time: " + (t1 - t0)/1e9d + "sec")
    result
  }

  def loadBusiness(jsonObj: JObject): (String, Array[String]) = {
    implicit val formats = DefaultFormats
    val catArr:ArrayBuffer[String] = ArrayBuffer()
    if ((jsonObj \ "categories") != JNull &&(jsonObj \ "categories") != JNothing) {
      if (!(jsonObj \ "categories").extract[String].contentEquals("")) {
        val categories = (jsonObj \ "categories").extract[String].split(",")
        for (category <- categories) {
          catArr += category.trim()
        }
      }
    }
    ((jsonObj \ "business_id").extract[String], catArr.toArray)
  }

  def loadReviews(jsonObj: JObject): (String, (Double, Int)) = {
    implicit val formats = DefaultFormats
    val businessId = (jsonObj \ "business_id").extract[String]
    val stars = (jsonObj \ "stars").extract[Double]
    (businessId, (stars, 1))
  }

  def mapCategory(tuple: (Array[String], (Double, Int))): Array[(String, (Double, Int))] = {
    val catArr: Array[String] = tuple._1
    val starsCount: (Double, Int) = tuple._2
    val catMap: ArrayBuffer[(String, (Double, Int))] = ArrayBuffer()
    for (value <- catArr) {
      catMap += ((value, starsCount))
    }
    catMap.toArray
  }

  type BusinessType = Map[String, Array[String]]
  type ReviewType = Map[String, (Double, Int)]
  def loadBusinessJson(businessFile: String): (BusinessType, ReviewType) = {
    implicit val formats = DefaultFormats
    var idCat = Map[String, Array[String]]()
    var catCount = Map[String, (Double, Int)]()

    val source = Source.fromFile(businessFile)
    for (line <- source.getLines()) {
      val jsonObj = parse(line)

      if ((jsonObj \"categories") != JNull && (jsonObj \ "categories") != JNothing) {
        if (!(jsonObj \ "categories").extract[String].contentEquals("")) {
          val categories = (jsonObj \ "categories").extract[String].split(",")
          val catArr: ArrayBuffer[String] = ArrayBuffer()
          for (category <- categories) {catArr += category.trim()}
          idCat += (jsonObj \ "business_id").extract[String] -> catArr.toArray

          for (category <- catArr) {
            if (!(catCount isDefinedAt category)) {
              catCount += category -> (0.0, 0)
            }
          }
        }
      }
    }
    source.close()
    (idCat, catCount)
  }

  def mapBusinessReviews(reviewFile: String, idCat: task2.BusinessType, catCount: task2.ReviewType): scala.collection.mutable.Map[String, (Double, Int)] = {
    implicit val formats = DefaultFormats
    var newCatCount = scala.collection.mutable.Map[String, (Double, Int)]()

    val source = Source.fromFile(reviewFile)
    for (line <- source.getLines()) {

      val jsonObj = parse(line)

      if (idCat isDefinedAt (jsonObj \ "business_id").extract[String]) {
        val catArr: Array[String] = idCat((jsonObj \ "business_id").extract[String])

        for (category <- catArr) {
          var star: Double = if (newCatCount.contains(category)) newCatCount(category)._1 else catCount(category)._1
          star += (jsonObj \ "stars").extract[Double]

          var count: Int = if (newCatCount.contains(category)) newCatCount(category)._2 else catCount(category)._2
          count += 1

          newCatCount += category -> (star, count)
        }
      }
    }
    source.close()
    newCatCount
  }

  def sortTopAvg(results: Array[(String, Double)], n: Int): String = {
    var avgMap = Map[Double, Array[String]]()

    for (i <- results) {
      if (!(avgMap isDefinedAt i._2)) {
        avgMap += (i._2 -> Array(i._1))
      } else {
        val tempArr: Array[String] = avgMap(i._2)
        val tempBuffer = Array(i._1)
        avgMap += (i._2 -> concat(tempArr, tempBuffer))
      }
    }

    //Resort map by key
    val avgMapSorted = ListMap(avgMap.toSeq.sortWith(_._1 > _._1):_*)
    var count = n

    var str = "["

    breakable {
      for ((k, v) <- avgMapSorted) {
        if (count == 0) {
          break
        }
        quickSort(v) //sort array
        for (i <- v) {
          if (count == 0) {
            break
          }
          str += "[\"" + i + "\", " + k + "], "
          count -= 1
        }
      }
    }
    str = str.substring(0, str.length - 2)
    str += "]"
    str
  }

  def sparkImplementation(reviewFile: String, businessFile: String, n: Int): String = {
    Logger.getLogger("org").setLevel(Level.ERROR)

    val sc = new SparkContext("local[*]", "Task2")

    val businessJson = sc.textFile(businessFile)
    val reviewJson = sc.textFile(reviewFile)

    val businessRDD = businessJson.map(x => parse(x).asInstanceOf[JObject]).map(loadBusiness)
    val reviewsRDD = reviewJson.map(x => parse(x).asInstanceOf[JObject]).map(loadReviews)

    // Natural Join Business and Reviews
    val businessReviews = businessRDD.join(reviewsRDD).cache()

    val categoryStars = businessReviews.filter(x => x._2._1.length != 0)
      .map(x => mapCategory(x._2))
      .flatMap(x => x)
      .reduceByKey((x, y) => (x._1 + y._1, x._2 + y._2)).cache()

    // Compute the average
    val avgStars = categoryStars.mapValues(x => x._1/x._2).sortBy(_._2, ascending = false)
    val results = avgStars.collect()

    val outDict = "result" -> parse(sortTopAvg(results, n))
    compact(render(outDict))
  }

  def implementation(reviewFile: String, businessFile: String, n: Int): String = {
    val (idCat, catCount) = loadBusinessJson(businessFile)

    val joinedCatCount = mapBusinessReviews(reviewFile, idCat, catCount)

    var avgCatDict = Map[String, Double]()

    for ((k, v) <- joinedCatCount) {
      if (v._2 == 0) {
        avgCatDict += k -> 0.0
      } else {
        avgCatDict += k -> v._1/v._2
      }
    }

    //Sort Average Count Dictionary by average value in Descending order
    val avgCatDictSorted = ListMap(avgCatDict.toSeq.sortWith(_._2 > _._2):_*)

    //Convert Map[String, Double] to Array[(String, Double)]
    val avgCatArr: ArrayBuffer[(String, Double)] = ArrayBuffer()

    for ((k, v) <- avgCatDictSorted) {
      avgCatArr += ((k, v))
    }

    val results: String = sortTopAvg(avgCatArr.toArray, n)

    val outDict = "result" -> parse(results)
    compact(render(outDict))
  }


  def main(args: Array[String]): Unit = {

    // Unpack Arguments
    val Array(reviewFile, businessFile, outputFile, ifSpark, n) = args

    var result = ""

    if (ifSpark.contentEquals("spark")) {
      result = time(sparkImplementation(reviewFile, businessFile, n.toInt))
    } else {
      result = time(implementation(reviewFile, businessFile, n.toInt))
    }

    val writer = new FileWriter(outputFile)
    writer.write(result)
    writer.close()
  }
}
